apiserver/facades/client/secrets/secrets.go

```diff
@@ -429,29 +437,44 @@ func (s *SecretsAPI) updateSecret(backend provider.SecretsBackend, arg params.Up
 		return errors.Trace(err)
 	}
 	if len(arg.Content.Data) > 0 {
-		revId, err := backend.SaveContent(context.TODO(), uri, md.LatestRevision+1, coresecrets.NewSecretValue(arg.Content.Data))
-		if err != nil && !errors.Is(err, errors.NotSupported) {
-			return errors.Trace(err)
+		v := coresecrets.NewSecretValue(arg.Content.Data)
+		checksum, err := v.Checksum()
+		if err != nil {
+			return errors.Annotate(err, "calculating secret checksum")
 		}
-		if err == nil {
-			defer func() {
-				if errOut != nil {
-					// If we failed to update the secret, we should delete the
-					// secret value from the backend for the new revision.
-					if err2 := backend.DeleteContent(context.TODO(), revId); err2 != nil &&
-						!errors.Is(err2, errors.NotSupported) &&
-						!errors.Is(err2, errors.NotFound) {
-						logger.Errorf("failed to delete secret %q: %v", revId, err2)
+		if checksum == md.LatestRevisionChecksum {
```

Comment:
Starting to feel like this method is doing way too much. Not an ask for here, but when you go to integrate this into 4.0, can you please look to improve on this? (it will probably happen naturally anyway with the business logic moving into the service).

Reply:
Yes, it's all different with domain services. Our facades in 3.x are quite bloated because that's where loo many abstractions are blurred together.